pkg/controllers/cloud/account_poller.go

```diff
@@ -428,22 +428,42 @@ func (p *accountPoller) getVMSelectorMatch(vm *cloudv1alpha1.VirtualMachine) *cl
 	vmSelectors, _ := p.vmSelector.ByIndex(virtualMachineSelectorMatchIndexerByID, vm.Annotations[common.AnnotationCloudAssignedIDKey])
 	for _, i := range vmSelectors {
 		vmSelector := i.(*cloudv1alpha1.VirtualMachineSelector)
-		if vmSelector.Agented {
-			return vmSelector
+		return vmSelector
+	}
+
+	// VM Name is not unique, hence iterate all selectors matching the VM Name to see the best matching selector.
+	// VM intended to match a selector with vpcMatch and vmMatch selector, falls under exactMatchSelector.
+	// VM intended to match a selector with only vmMatch selector, falls under partialMatchSelector.
+	var exactMatchSelector *cloudv1alpha1.VirtualMachineSelector = nil
+	var partialMatchSelector *cloudv1alpha1.VirtualMachineSelector = nil
+	vmSelectors, _ = p.vmSelector.ByIndex(virtualMachineSelectorMatchIndexerByName, vm.Annotations[common.AnnotationCloudAssignedNameKey])
+	for _, i := range vmSelectors {
+		vmSelector := i.(*cloudv1alpha1.VirtualMachineSelector)
+		if vmSelector.VpcMatch != nil {
+			if vmSelector.VpcMatch.MatchID == vm.Annotations[common.AnnotationCloudAssignedVPCIDKey] {
+				exactMatchSelector = vmSelector
```

Comment:
As there is an exact match here, why not return from here directly?

```suggestion
				return vmSelector
```

Reply:
Done